allennlp/training/supervised_trainer.py

```diff
@@ -0,0 +1,816 @@
+
+import logging
+import os
+import shutil
+import time
+import re
+import datetime
+import traceback
+from typing import Dict, Optional, List, Tuple, Union, Iterable, Any
+
+import torch
+import torch.optim.lr_scheduler
+
+from allennlp.common import Params
+from allennlp.common.checks import ConfigurationError
+from allennlp.common.util import (dump_metrics, gpu_memory_mb, parse_cuda_device, peak_memory_mb,
+                                  get_frozen_and_tunable_parameter_names)
+from allennlp.common.tqdm import Tqdm
+from allennlp.data.instance import Instance
+from allennlp.data.iterators.data_iterator import DataIterator
+from allennlp.data.vocabulary import Vocabulary
+from allennlp.models.model import Model
+from allennlp.nn import util
+from allennlp.training.learning_rate_schedulers import LearningRateScheduler
+from allennlp.training.optimizers import Optimizer
+from allennlp.training.trainer import Trainer
+from allennlp.training.util import time_to_str, move_optimizer_to_cuda, datasets_from_params
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+@Trainer.register("default")
```

Comment:
This should be registered with a name that matches the class semantics (`"single_task"`?), and set as the default implementation in the base class, instead of registered with the name "default".

Reply:
the reason I did it this way is that the existing trainer was already registered as "default", and I didn't want to break that, although it's hard to imagine that anyone was using that

Reply:
I see, ok, that makes sense.  If you want to be paranoid about that particular backward compatibility, we could just register this twice.